File: sdks/python/beater/providers/anthropic.py
"""Drop-in Anthropic instrumentation: ``client = wrap_anthropic(Anthropic())``."""

from __future__ import annotations

from typing import Any

from opentelemetry.trace import Status, StatusCode

from ..observe import _apply_common, _to_value
from ..semconv import Attr, SpanKind
from ..tracing import get_tracer


def _record_usage(span, response: Any) -> None:
    usage = getattr(response, "usage", None)
    if usage is None:
        return
    input_tokens = getattr(usage, "input_tokens", None)
    output_tokens = getattr(usage, "output_tokens", None)
    cache_read = getattr(usage, "cache_read_input_tokens", None)
    if input_tokens is not None:
        span.set_attribute(Attr.LLM_TOKEN_PROMPT, int(input_tokens))
    if output_tokens is not None:
        span.set_attribute(Attr.LLM_TOKEN_COMPLETION, int(output_tokens))
    if cache_read is not None:
        span.set_attribute(Attr.LLM_TOKEN_CACHE_READ, int(cache_read))


def _output_text(response: Any) -> str:
    try:
        blocks = getattr(response, "content", None) or []
        parts = [getattr(block, "text", "") for block in blocks if getattr(block, "text", None)]
        return "".join(parts)
    except (AttributeError, TypeError):
        return ""


def wrap_anthropic(client: Any) -> Any:
    """Instrument an Anthropic client in place and return it.

    Wraps ``client.messages.create`` so each call emits an ``llm.call`` span with
    model and token counts. Safe to call once per client.
    """

    if getattr(client, "_beater_wrapped", False):
        return client

    messages_api = client.messages
    original = messages_api.create

    def create(*args: Any, **kwargs: Any) -> Any:
        tracer = get_tracer()
        model = kwargs.get("model", "unknown")
        with tracer.start_as_current_span("anthropic.messages.create") as span:
            _apply_common(span, SpanKind.LLM_CALL)
            span.set_attribute(Attr.LLM_PROVIDER, "anthropic")
            span.set_attribute(Attr.LLM_MODEL_NAME, str(model))
            messages = kwargs.get("messages")
            if messages is not None:
                span.set_attribute(Attr.INPUT_VALUE, _to_value(messages))
            try:
                response = original(*args, **kwargs)
                response_model = getattr(response, "model", None)
                if response_model:
                    span.set_attribute(Attr.LLM_MODEL_NAME, str(response_model))
                _record_usage(span, response)
                span.set_attribute(Attr.OUTPUT_VALUE, _output_text(response))
                span.set_status(Status(StatusCode.OK))
                return response
            except Exception as exc:  # noqa: BLE001
                span.set_status(Status(StatusCode.ERROR, str(exc)))
                span.record_exception(exc)
                raise

    messages_api.create = create  # type: ignore[method-assign]
    client._beater_wrapped = True
    return client

File: sdks/typescript/src/providers/anthropic.ts
/** Drop-in Anthropic instrumentation: `const client = wrapAnthropic(new Anthropic())`. */

import { context, Span, SpanStatusCode, trace } from "@opentelemetry/api";

import { Attr, SpanKind } from "../semconv";
import { getTracer } from "../tracing";
import { _internal } from "../observe";

const { toValue, applyCommon } = _internal;

/* eslint-disable @typescript-eslint/no-explicit-any */

const ANTHROPIC_CLIENT_MARK = Symbol.for("@beater/sdk.anthropic.client");
const ANTHROPIC_METHOD_MARK = Symbol.for("@beater/sdk.anthropic.method");

function recordUsage(span: any, response: any): void {
  const usage = response?.usage;
  if (!usage) return;
  if (usage.input_tokens != null) span.setAttribute(Attr.LLM_TOKEN_PROMPT, Number(usage.input_tokens));
  if (usage.output_tokens != null) span.setAttribute(Attr.LLM_TOKEN_COMPLETION, Number(usage.output_tokens));
  if (usage.cache_read_input_tokens != null) {
    span.setAttribute(Attr.LLM_TOKEN_CACHE_READ, Number(usage.cache_read_input_tokens));
  }
}

function outputText(response: any): string {
  try {
    const blocks = response.content ?? [];
    const text = blocks.map((b: any) => b?.text ?? "").join("");
    if (text) return text;
    const toolCalls = blocks
      .filter((b: any) => b?.type === "tool_use")
      .map((b: any) => ({ id: b?.id, name: b?.name, input: b?.input }));
    return toolCalls.length > 0 ? toValue({ tool_calls: toolCalls }) : "";
  } catch {
    return "";
  }
}

function isAsyncIterable(value: unknown): value is AsyncIterable<any> {
  return Boolean(value && typeof (value as any)[Symbol.asyncIterator] === "function");
}

interface StreamCapture {
  text: string[];
  toolCalls: Array<{ id?: string; name?: string; input?: unknown; partialJson?: string }>;
}

function recordStreamEvent(capture: StreamCapture, span: Span, event: any): void {
  const message = event?.message;
  if (message?.model) span.setAttribute(Attr.LLM_MODEL_NAME, String(message.model));
  recordUsage(span, message);
  recordUsage(span, event);

  if (event?.type === "content_block_start" && event.content_block?.type === "tool_use") {
    capture.toolCalls.push({
      id: event.content_block.id,
      name: event.content_block.name,
      input: event.content_block.input,
    });
  }

  if (event?.type === "content_block_delta") {
    if (event.delta?.text) capture.text.push(String(event.delta.text));
    if (event.delta?.partial_json) {
      const current = capture.toolCalls[capture.toolCalls.length - 1];
      if (current) current.partialJson = `${current.partialJson ?? ""}${event.delta.partial_json}`;
    }
  }
}

function finishSpan(span: Span, response: any): void {
  if (response?.model) span.setAttribute(Attr.LLM_MODEL_NAME, String(response.model));
  recordUsage(span, response);
  const output = outputText(response);
  if (output) span.setAttribute(Attr.OUTPUT_VALUE, output);
  span.setStatus({ code: SpanStatusCode.OK });
  span.end();
}

function finishStream(span: Span, capture: StreamCapture): void {
  const text = capture.text.join("");
  if (text) {
    span.setAttribute(Attr.OUTPUT_VALUE, text);
  } else if (capture.toolCalls.length > 0) {
    span.setAttribute(Attr.OUTPUT_VALUE, toValue({ tool_calls: capture.toolCalls }));
  }
  span.setStatus({ code: SpanStatusCode.OK });
  span.end();
}

function recordError(span: Span, err: unknown): void {
  span.setStatus({ code: SpanStatusCode.ERROR, message: String(err) });
  span.recordException(err as Error);
  span.end();
}

async function* wrapStream(stream: AsyncIterable<any>, span: Span): AsyncIterable<any> {
  const capture: StreamCapture = { text: [], toolCalls: [] };
  let failed = false;
  try {
    for await (const event of stream) {
      recordStreamEvent(capture, span, event);
      yield event;
    }
  } catch (err) {
    failed = true;
    recordError(span, err);
    throw err;
  } finally {
    if (!failed) finishStream(span, capture);
  }
}

function patchCreate(owner: any): boolean {
  if (!owner?.create || owner.create[ANTHROPIC_METHOD_MARK]) return false;
  const original = owner.create;

  owner.create = async function create(...args: any[]): Promise<any> {
    const params = args[0] ?? {};
    const tracer = getTracer();
    const span = tracer.startSpan("anthropic.messages.create");
    applyCommon(span, SpanKind.LLM_CALL);
    span.setAttribute(Attr.LLM_PROVIDER, "anthropic");
    span.setAttribute(Attr.LLM_MODEL_NAME, String(params.model ?? "unknown"));
    if (params.messages) span.setAttribute(Attr.INPUT_VALUE, toValue(params.messages));

    try {
      const active = trace.setSpan(context.active(), span);
      const response = await context.with(active, () => original.apply(this, args));
      if (isAsyncIterable(response)) {
        return wrapStream(response, span);
      }
      finishSpan(span, response);
      return response;
    } catch (err) {
      recordError(span, err);
      throw err;
    }
  };
  owner.create[ANTHROPIC_METHOD_MARK] = true;
  return true;
}

/** Instrument an Anthropic client in place and return it. */
export function wrapAnthropic<T extends Record<string, any>>(client: T): T {
  if ((client as any)[ANTHROPIC_CLIENT_MARK]) return client;
  patchCreate(client.messages);

  (client as any)[ANTHROPIC_CLIENT_MARK] = true;
  return client;
}
